design-documents/frontend/technical-vision.md

```diff
@@ -0,0 +1,140 @@
+
+_Disclaimer: though possible to introduce [TypeScript][] into an existing [CRA][] app and though [TypeScript][] can co-exist with JavaScript, it may not make sense to introduce it into an existing project if you feel it may hinder momentum._
+
+### Technologies
+- [Adding TypeScript to CRA](https://facebook.github.io/create-react-app/docs/adding-typescript).
+- Code **consistency** is enforced via [configuration](http://www.typescriptlang.org/docs/handbook/tsconfig-json.html) and [linting with TSLint](https://palantir.github.io/tslint/).
+   - _Disclaimer: use the most strict rules before you write a single line of code, because it's much harder to introduce these rules later._
+- Generate [TypeScript][] declarations for [CSS Modules](https://github.com/css-modules/css-modules) via [css-modules-typescript-loader](https://www.npmjs.com/package/css-modules-typescript-loader).
+- Documentation via [TSDoc](https://github.com/Microsoft/tsdoc#tsdoc).
+
+## Portability
+
+We should prepare to achieve as much code sharing as possible, regardless of environment. JavaScript is the only language that can run natively in all browsers; yet, it also [transcends the browser](https://cdb.reacttraining.com/universal-javascript-4761051b7ae9), running on servers and [even native mobile applications](https://facebook.github.io/react-native/). As such, the golden path for sharing code starts with JavaScript. By writing components in a decoupled way, we don't limit ourselves as to where those components can run.
+
+On top of sharing code between different environments, we should also be concerned about sharing code between projects. For this, we should introduce a business logic repository that includes common helper functions – logic that we're likely to use between projects.
+
+### Business Logic Repository
+1. Written and [documented](https://github.com/Microsoft/tsdoc#tsdoc) in [TypeScript][].
+1. [Semantic versioning](https://semver.org/) to ensure that progress can be made without breaking things for existing - consumers.
+1. 100% unit test coverage via [Jest][] for **consistency**.
+
+For **performance** reasons, JavaScript should only be used for cheap operations on the front end, reserving expensive operations for a services layer. It's worth calling out that we could achieve even more code sharing if this API were written in JavaScript. For practical reasons, however, it makes more sense to leverage internal PHP resources for this task.
+
+## Browser Support
+All users should have a **consistent** user experience. For the best **performance**, those with the latest browsers should not be penalized by incurring an additional payload hit than those with older browsers. To achieve this goal, we download just the [polyfills](https://remysharp.com/2010/10/08/what-is-a-polyfill) needed for our sites, tailored to each browser via [Polyfill.io](https://polyfill.io) feature detection.

```

Comment:
Since we haven't used `Polyfill.io` in a product yet, I'm hesitant to provide an explicit requirement for it. I think we can leave that in as a possible route to accomplish the goal, but just be more clear about the real intent here. Regarding this line:

> those with the latest browsers should not be penalized by incurring an additional payload hit than those with older browsers

I think this won't be clear to those unfamiliar with transpilation/polyfills _why_ the extra hit for modern browsers would happen.

Reply:
I actually learned recently that we (Adobe) don't "trust" polyfill.io:

> polyfill.io won’t detect usage, so it’ll just download all polyfills for your target browser, even if you don’t use like 90% of them.

☝️ A bit out of place, as you can tweak your Polyfill.io request to request only the features you use.

> But I’d argue the performance hit from the additional bandwidth is less than the perf hit from having an additional connection to polyfill.io (which will also break your website if it goes down!)

☝️ Good argument, but a bit subjective on the first point.

> If polyfill.io is loading external scripts, I think we’d run into PCI compliance issues.

☝️ That's something with which I can't really argue.

In conclusion, I'll generalize this point instead of the specific usage of polyfill.io.

Reply:
Addressed in 4f625a47b50e0c0ac67ccbf5fed03eff17b94d40